activerecord/test/cases/adapters/mysql2/sanitize_test.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require "cases/helper"
+require "models/post"
+
+class Mysql2SanitizeTest < ActiveRecord::Mysql2TestCase
+  def test_sanitize_sql_for_conditions
+    actual = ActiveRecord::Base.sanitize_sql_for_conditions(["name=? and group_id=?", "foo'bar", 4])
+    expected = "name='foo\\'bar' and group_id='4'"
+    assert_equal expected, actual
+
+    actual = ActiveRecord::Base.sanitize_sql_for_conditions(["name=:name and group_id=:group_id", name: "foo'bar", group_id: 4])
+    expected = "name='foo\\'bar' and group_id='4'"
+    assert_equal expected, actual
+
+    actual = ActiveRecord::Base.sanitize_sql_for_conditions(["name='%s' and group_id='%s'", "foo'bar", 4])
+    expected = "name='foo\\'bar' and group_id='4'"
+    assert_equal expected, actual
+
+    actual = ActiveRecord::Base.sanitize_sql_for_conditions("name='foo''bar' and group_id='4'")
+    expected = "name='foo''bar' and group_id='4'"
+    assert_equal expected, actual
+  end
+
+  def test_sanitize_sql_for_assignment
+    actual = Post.sanitize_sql_for_assignment(["name=? and group_id=?", nil, 4])
+    expected = "name=NULL and group_id='4'"
+    assert_equal expected, actual
+
+    actual = Post.sanitize_sql_for_assignment(["name=:name and group_id=:group_id", name: nil, group_id: 4])
+    expected = "name=NULL and group_id='4'"
+    assert_equal expected, actual
+
+    actual = Post.sanitize_sql_for_assignment({ name: nil, group_id: 4 })
+    expected = "`posts`.`name` = NULL, `posts`.`group_id` = 4"
+    assert_equal expected, actual
+
+    actual = Post.sanitize_sql_for_assignment("name=NULL and group_id='4'")
+    expected = "name=NULL and group_id='4'"
+    assert_equal expected, actual
+  end
+
+  def test_sanitize_sql_for_order
+    actual = Post.sanitize_sql_for_order([Arel.sql("field(id, ?)"), [1, 3, 2]])
+    expected = "field(id, '1','3','2')" # this is invalid SQL but expected behaviour
```

Comment:
This is not ideal. But it _is_ what is generated in the example in the docs.

Reply:
:thinking: For example `Post.sanitize_sql_for_order([Arel.sql("COALESCE(id, ?)"), [1, 3, 2]])` seems valid SQL fragment (tested with PostgreSQL). Why is that considered invalid SQL?

```ruby
irb(main):005:0> Post.order(Post.sanitize_sql_for_order([Arel.sql("COALESCE(id, ?)"), [1, 3, 2]]))
  Post Load (0.3ms)  SELECT "posts".* FROM "posts" ORDER BY COALESCE(id, 1,3,2)
```

Reply:
You are right, this is a bad example.

The original described bug was:
```ruby
ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: 12, per_page: 32 }])
```

Expected to produce `"LIMIT 12, 32"` but produced `"LIMIT '12', '32'"`.

But **specifically in the MySQL adapter**. 
```sql
mysql> SELECT * FROM examples LIMIT '1', '2';
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''1', '2'' at line 1
```

What has been hard about writing good docs and tests is that the output here differs on the adapter implementation, because how we handle quoting depends on the adapter selected in the current context.

I can update this test and comment for accuracy and clarity.

Reply:
Yes. I can do that. But also I was trying to assert what we have in the docs matches actual behaviour. Since we don't specify in the docs that the examples come from a specific adapter. At least now I updated them to explain that behaviour depends on the DB and adapter in use.